go-controller/pkg/ovn/controller/network_qos/network_qos_node.go

```diff
@@ -0,0 +1,160 @@
+package networkqos
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+func (c *Controller) processNextNQOSNodeWorkItem(wg *sync.WaitGroup) bool {
+	wg.Add(1)
+	defer wg.Done()
+	nqosNodeKey, quit := c.nqosNodeQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.nqosNodeQueue.Done(nqosNodeKey)
+	err := c.syncNetworkQoSNode(nqosNodeKey)
+	if err == nil {
+		c.nqosNodeQueue.Forget(nqosNodeKey)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with: %v", nqosNodeKey, err))
+
+	if c.nqosNodeQueue.NumRequeues(nqosNodeKey) < maxRetries {
+		c.nqosNodeQueue.AddRateLimited(nqosNodeKey)
+		return true
+	}
+
+	c.nqosNodeQueue.Forget(nqosNodeKey)
+	return true
+}
+
+// syncNetworkQoSNode decides the main logic everytime
+// we dequeue a key from the nqosNodeQueue cache
+func (c *Controller) syncNetworkQoSNode(key string) error {
+	startTime := time.Now()
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	klog.V(5).Infof("Processing sync for Node %s in Network QoS controller", name)
+
+	defer func() {
+		klog.V(5).Infof("Finished syncing Node %s Network QoS controller: took %v", name, time.Since(startTime))
+	}()
+	node, err := c.nqosNodeLister.Get(name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if !c.isNodeInLocalZone(node) && c.TopologyType() == types.Layer3Topology {
+		// clean up qos/address set for the node
+		return c.cleanupQoSFromNode(node.Name)
+	}
+	// configure qos for pods on the node
+	pods, err := c.getPodsByNode(node.Name)
+	if err != nil {
+		return err
+	}
+	switchName := c.getLogicalSwitchName(node.Name)
+	_, err = c.findLogicalSwitch(switchName)
+	if err != nil {
+		klog.V(4).Infof("Failed to look up logical switch %s: %v", switchName, err)
+		return err
+	}
+	for _, cachedKey := range c.nqosCache.GetKeys() {
+		err := c.nqosCache.DoWithLock(cachedKey, func(nqosKey string) error {
+			nqosObj, _ := c.nqosCache.Load(nqosKey)
+			if nqosObj == nil {
+				klog.Warningf("NetworkQoS not synced yet: %s", nqosKey)
+				// requeue nqos key to sync it
+				c.nqosQueue.Add(nqosKey)
+				// requeue namespace key 3 seconds later, allow NetworkQoS to be handled
+				c.nqosNamespaceQueue.AddAfter(key, 3*time.Second)
```

Comment:
why does it queue after 3 seconds? this seems like it could become a race. Did you copy this from ANP or something? I think it would be better to queue nQOS keys in other handlers, and then do everything in that thread.

Reply:
the situation here is that the NetworkQoS may not get reconciled yet, so line 82 adds the key to trigger a sync for it. if we requeue namespace key here immediately, namespace handler probably still won't be able to see NetworkQoS in cache, so want to add 3 seconds delay so that NetworkQoS handler will have time to process it.